### DISCUSSION TITLE
Defining persistent functions

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [mdgrs-mei/PowerShellRun#69] [Ideas]
Currently, we can define functions like this:

```powershell
Start-PSRunFunctionRegistration
Function global:LuckyNumber {
    Param([int]$Count = 1)
    Get-Random -Minimum 1 -Maximum 1000 -Count $count
}
Stop-PSRunFunctionRegistration
```

The problem is that if you re-run your configuration script *in the current session*, this function, `LuckyNumber` won't be displayed. Although it exist and you can run it from a prompt.

After digging through your code, I think I have a solution. At the *end* of the `StopRegistration()` class method, save the currently defined functions.

```powershell
$script:DefinedFunctions = $this.entries.Name
```

In the `StartRegistration()` method, remove previously defined functions.

```powershell
[void] StartRegistration($errorAction) {
    if (-not ($null -eq $this.functionsAtRegisterStart)) {
        Write-Error -Message 'Function registration already started.' -Category InvalidOperation -ErrorAction $errorAction
        return
    }

    if ($script:DefinedFunctions.count -gt 0) {
        $script:DefinedFunctions | foreach-object {
            #Write-Host "Removing function:\$_" -ForegroundColor magenta
            Remove-Item "function:\$_"
        }
    }
    $this.functionsAtRegisterStart = (Get-Command -Type Function).Name
}
```

This way, when they are defined in the configuration file, they will be re-detected as "new".

So far, this seems to solve the problem. But I don't know if there are any other implications with this approach.

**Comment by mdgrs-mei** [comment]
I took a slightly different approach but the issue was already fixed in #62. I compared the script blocks between Start and Stop, not just names.

**Reply by jdhitsolutions** [reply]
Ok. I'll wait to test on the next release.

**Reply by mdgrs-mei** [reply]
The issue has been fixed in the v0.9.0 release. `Add-PSRunFunction` was also added.
